server/src/main/java/org/opensearch/common/util/FeatureFlags.java

```diff
@@ -111,86 +125,242 @@ public class FeatureFlags {
+        FeatureFlagsImpl() {
+            initFromDefaults();
+            initFromSysProperties();
+        }
+
+        /**
+         * Initialize feature flags map from the following sources:
+         * (Each source overwrites previous feature flags)
+         * - Set from setting default
+         * - Set from JVM system property if flag exists
+         * - Set from provided settings if flag exists
+         * @param openSearchSettings The settings stored in opensearch.yml.
+         */
+        void initializeFeatureFlags(Settings openSearchSettings) {
+            initFromDefaults();
+            initFromSysProperties();
+            initFromSettings(openSearchSettings);
+        }
+
+        /**
+         * Set all feature flags according to setting defaults.
+         * Overwrites existing entries in feature flags map.
+         * Skips flags which are locked according to TestUtils.FlagLock.
+         */
+        private void initFromDefaults() {
+            for (Setting<Boolean> ff : featureFlags.keySet()) {
+                if (TestUtils.FlagLock.isLocked(ff.getKey())) continue;
+                featureFlags.put(ff, ff.getDefault(Settings.EMPTY));
+            }
+        }
+
+        /**
+         * Update feature flags according to JVM system properties.
+         * Feature flags are true if system property is set as "true" (case-insensitive). Else feature set to false.
+         * Overwrites existing value if system property exists.
+         * Skips flags which are locked according to TestUtils.FlagLock.
+         */
+        private void initFromSysProperties() {
+            for (Setting<Boolean> ff : featureFlags.keySet()) {
+                if (TestUtils.FlagLock.isLocked(ff.getKey())) continue;
+                String prop = System.getProperty(ff.getKey());
+                if (prop != null) {
+                    featureFlags.put(ff, Boolean.valueOf(prop));
+                }
+            }
+        }
+
+        /**
+         * @param featureFlagName feature flag setting key
+         * @return true if feature enabled - else false
+         */
+        boolean isEnabled(String featureFlagName) {
+            for (Setting<Boolean> ff : featureFlags.keySet()) {
+                if (ff.getKey().equals(featureFlagName)) return featureFlags.get(ff);
+            }
+            return false;
+        }
+
+        /**
+         * @param ff feature flag setting
+         * @return true if feature enabled - else false
+         */
+        boolean isEnabled(Setting<Boolean> ff) {
+            if (!featureFlags.containsKey(ff)) return false;
+            return featureFlags.get(ff);
+        }
+
+        /**
+         * @param featureFlagName feature flag key to set
+         * @param value value for flag
+         */
+        void set(String featureFlagName, Boolean value) {
+            for (Setting<Boolean> ff : featureFlags.keySet()) {
+                if (ff.getKey().equals(featureFlagName)) featureFlags.put(ff, value);
```

Comment:
Initially, I was confused why we need to iterate over the keys for putting featureFlag, then I noticed need to compare the setting name, within Setting object that is the key. I am assuming this is done only during the initialization?